modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -3839,7 +3969,14 @@ private void finalizePartitionCounters() {
                 cctx.kernalContext().getSystemExecutorService(),
                 nonLocalCacheGroups(),
                 grp -> {
-                    grp.topology().finalizeUpdateCounters();
+                    AffinityTopologyVersion topVer = sharedContext().exchange().readyAffinityVersion();
+
+                    // Failed node's primary partitions or just all local backups in case of possible exchange merge.
+                    Set<Integer> parts = nodeId != null ?
```

Comment:
It this intended as an optimization ?
This is not really needed, because here all possible updates are received and all remaining gaps could be safely removed.